File: src/main/java/spring/springserver/global/data/BaseResponse.java
package spring.springserver.global.data;

import com.fasterxml.jackson.annotation.JsonInclude;
import spring.springserver.global.exception.status_code.StatusCode;
import org.springframework.http.HttpStatus;

import java.util.Map;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record BaseResponse<T> (
    int status,
    T data,
    ErrorResponse error
) {

  public static <T> BaseResponse<T> ok(T data) {

    return new BaseResponse<>(HttpStatus.OK.value(), data, null);
  }

  public static BaseResponse<Void> error(HttpStatus status, String code, String message) {

    return new BaseResponse<>(status.value(), null, ErrorResponse.of(code, message));
  }

  public static BaseResponse<Void> error(HttpStatus status, String code, String message, Map<String, String> details) {

    return new BaseResponse<>(status.value(), null, ErrorResponse.of(code, message, details));
  }

  public static BaseResponse<Void> error(HttpStatus status, ErrorResponse error) {

    return new BaseResponse<>(status.value(), null, error);
  }

  public static <T> BaseResponse<T> error(StatusCode status) {

    return new BaseResponse<>(status.getHttpStatus().value(), null, ErrorResponse.of(status.getCode(), status.getMessage()));
  }
}

File: src/main/java/spring/springserver/global/data/BaseResponse.kt
package spring.springserver.global.data

import com.fasterxml.jackson.annotation.JsonInclude
import org.springframework.http.HttpStatus

@JsonInclude(JsonInclude.Include.NON_NULL)
data class BaseResponse<T>(
    val status: Int,
    val data: T?,
    val error: ErrorResponse?
) {
    companion object {

        fun <T> ok(
            data: T
        ): BaseResponse<T> {

            return BaseResponse(
                status = HttpStatus.OK.value(),
                data = data,
                error = null
            )
        }

        fun ok(data: Nothing?): BaseResponse<Void> {

            return BaseResponse(
                status = HttpStatus.OK.value(),
                data = null,
                error = null
            )
        }

        fun error(
            httpStatus: HttpStatus,
            errorResponse: ErrorResponse
        ): BaseResponse<Void> {

            return BaseResponse(
                status = httpStatus.value(),
                data = null,
                error = errorResponse
            )
        }
    }
}
